### PR TITLE
WIP: Fix the issues with corner cases in Loop APIs

### PR DESCRIPTION
**Pull request by ChengJin01**
The changes are mainly focusing on a few corner cases
that were never addressed in the existing Loop APIs.

Signed-off-by: CHENGJin <jincheng@ca.ibm.com>

### CODE REVIEW

**Comment by ChengJin01** [comment]
Peer reviewer: @tajila
Senior reviewer: @DanHeidinga

**Comment by ChengJin01** [comment]
The change mainly addresses the following corner cases plus a few bugs previous not covered against Java 9 API Spec:
1) whileLoop: The parameter list (V A...) of the body constrain the parameter lists of the other loop parts,
              which means the external parameter of the body (A...) must be non-empty and longer than
              other argument handles.
2) countedLoop: If the external parameters of the body exist, it must be longer than other argument handles;
                if not, the end handle's parameters must be longer than other argument handles.
3) iteratedLoop: 
                 [1] fixed bugs with the first parameter & return type of iterator
                 [2] fixed a bug with the indexes of parameters passed to permuteArguments()
                 [3] The main constraints for the external parameter types (A...) are as follows:
		                 (1) if the external parameter types of body exist, it constrains the parameter types of other handles.
		                 (2) if not, the parameter types of iterator uniquely determines the external parameter types.
		                 (3) if iterator is null, the first external parameter type A1 must be Iterable or a subtype.

4) loop: 
         [1] a clause with all null handles should be ignored.
         [2] a little optimized code with iterationVarTypesArray
         [3] validate the iteration types of non-init handles after filling up all parameters to ensure
             all non-init handles are checked.

JSR test cases (including newly added ones) & Java 9 JCK passed with the new changes.

**Comment by ChengJin01** [comment]
@theresa-m were assigned to fix loop methods at https://github.com/eclipse/openj9/pull/624. So the changes will be integrated to her changes in there.

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 1]
```diff
@@ -3721,6 +3797,37 @@ public static MethodHandle iteratedLoop(MethodHandle iteratorHandle, MethodHandl
...
+		 */
+		if ((void.class == bodyReturnType) && (paramTypesLength > 1)) {
+			externalParamType = bodyParamTypes[1];
+		} else if ((void.class != bodyReturnType) && (paramTypesLength > 2)) {
+			externalParamType = bodyParamTypes[2];
+		}
+		
+		if (null == externalParamType) {
+			if (null != initHandle) {
+				Class<?>[]  initParamTypes = initHandle.type.arguments;
```
Comment (DanHeidinga) [inline jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: nitpick: spacing

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 2]
```diff
@@ -3740,42 +3847,44 @@ private static final void validateArgumentsOfIteratedLoop(MethodHandle iteratorH
...
 			}
 		}
 		
+		/* The main constraints for the external parameter types (A...) are as follows:
+		 * 1) if the external parameter types of body exist, it constrains the parameter types of other handles.
+		 * 2) if not, the parameter types of iterator uniquely determines the external parameter types.
+		 * 3) if iterator is null, the first external parameter type A1 must be Iterable or a subtype.
+		 */
+		
+		Class<?>  bodyExternalParamType = null;
```
Comment (DanHeidinga) [inline jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: Can this reuse the new `getLeadingExternalParamType` function?

File: jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java [hunk 3]
```diff
@@ -3868,16 +3980,25 @@ private static MethodHandle recreateIteratedBodyHandle(MethodHandle iteratorHand
...
-			Class<?> paramType2 = bodyParamTypes[1];
-			bodyParamTypes[0] = paramType2;
-			bodyParamTypes[1] = paramType1;
-			MethodType newBodyType = MethodType.methodType(bodyReturnType, bodyParamTypes);
-			loopBody = MethodHandles.permuteArguments(loopBody, newBodyType, 1, 0);
+			Class<?>[] oldBodyParamTypes = bodyType.arguments;
+			int bodyParamTypesLength = oldBodyParamTypes.length;
+			Class<?>[] newBodyParamTypes = oldBodyParamTypes.clone();
+			int[] bodyParamTypesIndexes = new int[bodyParamTypesLength];
+			for (int paramTypeIndex = 0; paramTypeIndex < bodyParamTypesLength; paramTypeIndex++) {
```
Comment (DanHeidinga) [inline jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java]: Can this loop be replaced with a System.arraycopy?

**Comment by ChengJin01** [comment]
The change here will be abandoned once I finish the code review at https://github.com/eclipse/openj9/pull/624 (Theresa's changes for loop methods) and confirm all corner cases are correctly handled in there.

**Comment by DanHeidinga** [comment]
Marking this as WIP as @ChengJin01 indicates this PR will be abandoned.

**Pull request closed without merging**